Mathlib/AlgebraicGeometry/EllipticCurve/Modular/TateNormalForm.lean

```diff
@@ -0,0 +1,268 @@
+-/
+
+noncomputable section
+
+namespace WeierstrassCurve
+
+/-- The Tate normal form is $$y^2 + (1-c)xy - by = x^3 - bx^2$$. -/
+@[mk_iff]
+class IsTateNF {R : Type*} [Zero R] (W : WeierstrassCurve R) : Prop where
+  a₂₃ : W.a₂ = W.a₃
+  a₄  : W.a₄ = 0
+  a₆  : W.a₆ = 0
+
+namespace Affine.Point
+
+variable {R : Type*} [CommRing R] {W : WeierstrassCurve R} (P : W.toAffine.Point)
+
+/-- Typeclass for a given point not being zero (the point at infinity). -/
+@[mk_iff]
+class NeZero : Prop where
+  neZero : P ≠ 0
+
+instance : DecidablePred (Affine.Point.NeZero (W:=W))
+| zero   => isFalse (fun ⟨h⟩ ↦ h rfl)
+| some _ => isTrue ⟨Point.noConfusion⟩
+
+/-- The `X` coordinate of a given point. For the point at infinity, this returns `0`
+(junk value). -/
+def X : W.toAffine.Point → R
+  | 0 => 0
+  | @some _ _ _ x _ _ => x
+
+/-- The `Y` coordinate of a given point. For the point at infinity, this returns `0`
+(junk value). -/
+def Y : W.toAffine.Point → R
+  | 0 => 0
+  | @some _ _ _ _ y _ => y
+
+variable (W) in
+@[simp] lemma not_neZero_zero : ¬(NeZero (0 : W.toAffine.Point)) :=
+  fun ⟨h⟩ ↦ h rfl
+
+lemma equation_X_Y [NeZero P] : W.toAffine.Equation P.X P.Y := by
+  cases P with
+  | zero => exact (not_neZero_zero W).elim (by assumption)
+  | some h => exact h.1
+
+lemma equation_X_Y' [NeZero P] : P.Y^2 + W.a₁ * P.X * P.Y + W.a₃ * P.Y
+    = P.X^3 + W.a₂ * P.X^2 + W.a₄ * P.X + W.a₆ :=
+  (equation_iff ..).1 P.equation_X_Y
+
+/-- The partial derivative `∂W/∂X` of the Weierstrass cubic at a given point `P`. -/
+def pX : R :=
+  W.a₁ * P.Y - (3 * P.X ^ 2 + 2 * W.a₂ * P.X + W.a₄)
+
+/-- The partial derivative `∂W/∂Y` of the Weierstrass cubic at a given point `P`. -/
+def pY : R :=
+  2 * P.Y + W.a₁ * P.X + W.a₃
+
+/-- The condition `2 • P ≠ 0` on all fibres. (TODO) -/
+@[mk_iff]
+class TwiceNeZero : Prop extends P.NeZero where
+  twiceNeZero : IsUnit P.pY
+
+lemma isUnit_pY [P.TwiceNeZero] : IsUnit P.pY :=
+  TwiceNeZero.twiceNeZero
+
+lemma pY_ne_zero [P.TwiceNeZero] [Nontrivial R] : P.pY ≠ 0 :=
+  P.isUnit_pY.ne_zero
+
+/-- The inverse of `pY` as a unit, whenever `2 • P ≠ 0` (i.e. `P.TwiceNeZero`). -/
+def pY_inv [P.TwiceNeZero] : Rˣ :=
+  P.isUnit_pY.unit⁻¹
+
+@[simp] lemma pY_mul_inv [P.TwiceNeZero] : P.pY * P.pY_inv = 1 := by
+  have : P.isUnit_pY.unit * P.isUnit_pY.unit⁻¹ = 1 := mul_inv_cancel _
+  rwa [Units.ext_iff, Units.val_mul, IsUnit.unit_spec] at this
+
+@[simp] lemma pY_inv_mul [P.TwiceNeZero] : P.pY_inv * P.pY = 1 :=
+  (mul_comm ..).trans P.pY_mul_inv
+
+@[simp] lemma pY_inv_inv [P.TwiceNeZero] : P.pY_inv⁻¹ = P.pY := by
+  rw [pY_inv, inv_inv]; rfl
+
+/-- A quantity that determines whether `3 • P = 0`. -/
+def μ [P.TwiceNeZero] : R :=
+  W.a₂ + 3 * P.X + P.pX * P.pY_inv * W.a₁ - (P.pX * P.pY_inv) ^ 2
+
+/-- The condition `3 • P ≠ 0` on all fibres. -/
+@[mk_iff]
+class ThriceNeZero : Prop extends P.NeZero where
+  thriceNeZero : IsUnit ((W.a₂ + 3 * P.X) * P.pY ^ 2 + P.pX * W.a₁ * P.pY - P.pX ^ 2)
```

Comment:
Does this already assume that you've made `a_6 = 0`? Do you want to make `3P \ne 0` an independent condition?